lib/cli/src/Cardano/CLI.hs

```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
   where
     preferences = prefs showHelpOnEmpty
 
+{-------------------------------------------------------------------------------
+                            Commands - 'HD Derivation'
+-------------------------------------------------------------------------------}
+
+cmdKey :: Mod CommandFields (IO ())
+cmdKey = command "key" $ info (helper <*> cmds) $ mempty
+    <> progDesc "Derive keys from mnemonics."
+  where
+    cmds = subparser $ mempty
+        <> cmdRootKey
+
+-- | Parse a key type
+--
+-- Note that we in the future might replace the type @ByronWalletStyle@ with
+-- another type, to include Shelley keys.
+parseKeyType :: Parser (IO ByronWalletStyle)
+parseKeyType = toIO . fromText <$> strOption
+    ( long "type"
+    <> metavar "KEYTYPE"
+    <> helpDoc (Just (vsep typeOptions))
+    )
+  where
+    toIO :: Either TextDecodingError ByronWalletStyle -> IO ByronWalletStyle
+    toIO (Right x) =
+        return x
+    toIO (Left (TextDecodingError e)) =
+        hPutStrLn stderr e >> exitFailure
+
+    typeOptions = string <$>
+        [ "Any of the following:"
+        , "  random (Daedalus, " ++ allowedWords Random ++ ")"
+        , "  icarus (" ++ allowedWords Icarus ++ ")"
+        , "  trezor (" ++ allowedWords Trezor ++ ")"
+        , "  ledger (" ++ allowedWords Ledger ++ ")"
+        ]
+
+    allowedWords = (++ " words")
+        . formatEnglishEnumeration
+        . map show
+        . allowedWordLengths
+        . byronCliKeyScheme
+       where
+          -- >>> formatEnglishEnumeration ["a", "b", "c"]
+          -- "a, b or c"
+          --
+          -- >>> formatEnglishEnumeration ["a", "b"]
+          -- "a or b"
+          --
+          -- >>> formatEnglishEnumeration ["a"]
+          -- "a"
+         formatEnglishEnumeration = formatEnglishEnumerationRev . reverse
+         formatEnglishEnumerationRev [ult, penult]
+            = penult ++ " or " ++ ult
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
+                ++ " or "
+                ++ ult
+         formatEnglishEnumerationRev xs = intercalate ", " (reverse xs)
+
+parseSeedWords :: Parser [Text]
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Lay thine eyes upon the type parameters and see that there are none.
+data CLIKeyScheme = CLIKeyScheme
+    { wordsToSeed :: [Text] -> Either String (Passphrase "seed")
+    , allowedWordLengths :: [Int]
+    , seedToRootKey :: Passphrase "seed" -> XPrv
+    }
+
+byronCliKeyScheme :: ByronWalletStyle -> CLIKeyScheme
+byronCliKeyScheme = \case
+    Random -> mkScheme (Proxy @'Random)
+            (Byron.getKey . flip Byron.generateKeyFromSeed pass)
```

Comment:
For consistency and readability with below, I'd be better to have that as a named function as well :+1:

Reply:
done.